cluster/partitioner/partitioner.go

```diff
@@ -12,41 +18,123 @@ type Partitioner interface {
 }
 
 type Kafka struct {
-	PartitionBy string
-	Partitioner sarama.Partitioner
+	PartitionBy     string
+	Partitioner     sarama.Partitioner
+	GetPartitionKey func(schema.PartitionedMetric, []byte) []byte
 }
 
 func NewKafka(partitionBy string) (*Kafka, error) {
+	kafka := Kafka{
+		PartitionBy: partitionBy,
+	}
+
 	switch partitionBy {
 	case "byOrg":
+		kafka.Partitioner = sarama.NewHashPartitioner("")
+		kafka.GetPartitionKey = func(m schema.PartitionedMetric, b []byte) []byte { return m.KeyByOrgId(b) }
 	case "bySeries":
+		kafka.Partitioner = sarama.NewHashPartitioner("")
+		kafka.GetPartitionKey = func(m schema.PartitionedMetric, b []byte) []byte { return m.KeyBySeries(b) }
+	case "bySeriesWithTags":
+		kafka.Partitioner = sarama.NewHashPartitioner("")
+		kafka.GetPartitionKey = func(m schema.PartitionedMetric, b []byte) []byte { return m.KeyBySeriesWithTags(b) }
 	default:
-		return nil, fmt.Errorf("partitionBy must be one of 'byOrg|bySeries'. got %s", partitionBy)
+		return nil, fmt.Errorf("partitionBy must be one of 'byOrg|bySeries|bySeriesWithTags'. got %s", partitionBy)
 	}
-	return &Kafka{
-		PartitionBy: partitionBy,
-		Partitioner: sarama.NewHashPartitioner(""),
-	}, nil
+
+	return &kafka, nil
 }
 
 func (k *Kafka) Partition(m schema.PartitionedMetric, numPartitions int32) (int32, error) {
-	key, err := k.GetPartitionKey(m, nil)
+	key := k.GetPartitionKey(m, nil)
+	return k.Partitioner.Partition(&sarama.ProducerMessage{Key: sarama.ByteEncoder(key)}, numPartitions)
+}
+
+type jumpPartitionerMauro struct{}
+
+func (p jumpPartitionerMauro) Partition(message *sarama.ProducerMessage, numPartitions int32) (int32, error) {
+	key, err := message.Key.Encode()
 	if err != nil {
 		return 0, err
 	}
-	return k.Partitioner.Partition(&sarama.ProducerMessage{Key: sarama.ByteEncoder(key)}, numPartitions)
+
+	// we need to get from a []byte key to a uint64 key, because jump requires that as input
+	// so we keep adding slices of 8 bytes to the jump key as uint64 values
+	// if the result wraps around the uin64 boundary it doesn't matter because
+	// this only needs to be fast and consistent
+	key = append(key, make([]byte, 8-len(key)%8)...)
+	var jumpKey uint64
+	for pos := 8; pos <= len(key); pos += 8 {
+		jumpKey += binary.BigEndian.Uint64(key[pos-8 : pos])
+	}
+
+	return jump.Hash(jumpKey, int(numPartitions)), nil
 }
 
-func (k *Kafka) GetPartitionKey(m schema.PartitionedMetric, b []byte) ([]byte, error) {
-	switch k.PartitionBy {
-	case "byOrg":
-		// partition by organisation: metrics for the same org should go to the same
-		// partition/MetricTank (optimize for locality~performance)
-		return m.KeyByOrgId(b), nil
-	case "bySeries":
-		// partition by series: metrics are distrubted across all metrictank instances
-		// to allow horizontal scalability
-		return m.KeyBySeries(b), nil
+func (p jumpPartitionerMauro) RequiresConsistency() bool {
```

Comment:
what is `RequiresConsistency()` for?  is this meant to be part of the "Partitioner" interface?

Reply:
it is part of the sarama.Partitioner interface.
https://github.com/Shopify/sarama/blob/a74fb772b5cfe121956313c4f6596e372e83cd27/partitioner.go#L17-L22